USB Emulation enhancement and fixes

fysnet opened the pull request:
My new code:
 Minor:
  1) Check speed indicator in the Transfer Descriptor matching the device speed.
  2) Max packet size check, checking that the amount requested/sent is within limits.
  3) Checking that the first packet sent is the GetDescriptor Request with a length equal to or less than the max packet size.
  4) Checking the command length in the USB SCSI emulation.
  5) Checking command parameters within various emulations (SCSI (BBB), etc.)
  6) Implemented "Boot Protocol" for the (HID) Mouse emulation.
  7) Fixed wheel mouse HID report (along with the other HID reports)
  8) Added the Over Current bits to the UHCI code
  9) Added the monitoring of the toggle bit in the TDs

 Major:
  1) Re-wrote the UHCI stack processing so that control and bulk reclamation can be used.
     (This made it so Windows 7 now works)
  2) Adding function to the USB SCSI emulation.
  3) Enhanced the xHCI emulation to further check for errors in a driver.

stlintel commented on bochs/iodev/usb/scsi_device.cc:
```diff
@@ -485,12 +485,12 @@ Bit8u* scsi_device_t::scsi_get_buf(Bit32u tag)
 }
 
-Bit32s scsi_device_t::scsi_send_command(Bit32u tag, Bit8u *buf, int lun, bool async)
+Bit32s scsi_device_t::scsi_send_command(Bit32u tag, Bit8u *buf, const Bit8u cmd_len, int lun, bool async)
```
why would you add const here ?
it doesn't matter as cmd_len is copied by value anyway

fysnet replied:
Sorry, habit I guess. Wanted to make sure that I didn't modify it throughout. You may remove it if you wish.

stlintel commented:
I would appreciate if next time you split style changes from functional changes

stlintel commented on bochs/iodev/usb/scsi_device.cc:
```diff
@@ -789,6 +797,166 @@ Bit32s scsi_device_t::scsi_send_command(Bit32u tag, Bit8u *buf, int lun, bool as
+              BX_DEBUG(("Event Status Notify: bit 0 in the request byte is set"));
+              Class = 0;
+            }
+            if (buf[4] & (1<<1)) { // Opertaional Change request bit is set
```
typo in comment
also I suggest to have nice enums instead of using magic const bitmasks like (1<<2)
these bits have names - you even mention them in comments

fysnet replied:
Indeed. You are right. My apologizes. Would you like a new pull request now, or may I wait until my next group of updates?

stlintel replied:
Please keep updating the same PR. This is the easiest way to handle things.

stlintel commented on bochs/iodev/usb/scsi_device.cc:
```diff
@@ -789,6 +797,166 @@ Bit32s scsi_device_t::scsi_send_command(Bit32u tag, Bit8u *buf, int lun, bool as
+              //p[0] = 0; // Bits 3:0 = Event Code (must be zero until an event occurs, then zero after read until next event occurs)
+            } else
+            if (buf[4] & (1<<2)) { // Power Management request bit is set
+              Class = 2;
```
I guess you could just determine class using smth like "lzcnt' primitive and remove most of the of statements here

fysnet replied:
I left it this way so that I could later add the remaining function. This addition was to pass a check that Win7 does in its USB enumeration of CDROM's. I intend to come back and add this function, though it isn't a high priority at the moment.

stlintel replied:
No problem

stlintel commented on bochs/iodev/usb/scsi_device.cc:
```diff
@@ -789,6 +797,166 @@ Bit32s scsi_device_t::scsi_send_command(Bit32u tag, Bit8u *buf, int lun, bool as
+            outbuf[ 4] = 1; // number of sessions (LSB)
+            outbuf[ 5] = 1; // first track number in last session (LSB)
+            outbuf[ 6] = 1; // last track number in last session (LSB)
+            outbuf[ 7] = (0 << 7) | (0 << 6) | (1 << 5) | (0 << 4) | (0 << 3)  | (0 << 2) | (0 << 0); // DID_V, DBC_V, URU, DAC_V, R, Legacy?, BG Format Status
```
again, suggest to have defines for all these bits
see examples in CPU code

fysnet replied:
Suggestion taken. Please see previous comments. I will take the time to add them.

stlintel commented on bochs/iodev/usb/usb_common.h:
```diff
@@ -67,31 +69,39 @@
 #define USB_DIR_IN   0x80
 
-#define USB_TYPE_MASK			(0x03 << 5)
+#define USB_TYPE_MASK			  (0x03 << 5)
```
is this space intentional ?
now code look misaligned

fysnet replied:
I saw that too. Found out that it is a difference in spaces and tabs used. The previous commit used tabs. When I modified the line (later reverting back to the original value), my compiler automatically changed the tab to spaces. I did not notice, since they still line up if you have two spaces for every tab. I only noticed afterwards with the submitted diff's. I always use spaces instead of tabs (just preference).

fysnet replied:
I have fixed, but will take a new PR since I have since made other changes to this file.

stlintel commented:
I had few comments about style - please try to address or let's discuss them.
I cant review functional part of code but here I trust  you as true code owner of all this stuff.
I also I would appreciate if next time you split style changes from functional changes.

stlintel commented:
This one looks me severe warning too:

```
In static member function 'static Bit64s usb_hub_device_c::hub_param_handler(bx_param_c*, bool, Bit64s)',
    inlined from 'static Bit64s usb_hub_device_c::hub_param_handler(bx_param_c*, bool, Bit64s)' at usb_hub.cc:719:8:
usb_hub.cc:744:16: warning: 'this' pointer is null [-Wnonnull]
  744 |       BX_PANIC(("hub_param_handler: external hub not found"));
../../logio.h:154:38: note: in definition of macro 'BX_PANIC'
  154 | #define BX_PANIC(x) (LOG_THIS panic) x
```

stlintel commented:
Will this PR fix (0xc0000417 crash while exiting with uhci enabled #25) ?

fysnet commented:
I apologize for the single commit of style and function changes. I was unaware that you preferred them separate. My intent is to not modify the style of anything outside the USB code since it is not of my origin. However, with the USB code, having a big part of its origin, I was less worried about style and more worried about function. However, I will keep your comments in mind with any further updates. Thank you.

fysnet commented:
Please keep me advise in the action you wish to take to commit these changes. Do you want a new pull request with the defines instead of bit reps now, or can they be a later pull request? Thank you for your time.

stlintel commented:
> Please keep me advise in the action you wish to take to commit these changes. Do you want a new pull request with the defines instead of bit reps now, or can they be a later pull request? Thank you for your time.

Please keep updating the same PR. This is the easiest way to handle things.

stlintel commented:
> I apologize for the single commit of style and function changes. I was unaware that you preferred them separate. My intent is to not modify the style of anything outside the USB code since it is not of my origin. However, with the USB code, having a big part of its origin, I was less worried about style and more worried about function. However, I will keep your comments in mind with any further updates. Thank you.

It is very hard to review 2000 lines changed over 26 files. It is huge code rewrite.
And then I notice that 90% of it just style changes.
They could be pushed almost immediately - I could just go over all lines and verify that nothing functional was affected.
I am fine with style changes that you do for you - so this would go quick.
Also in case of a problem in the code it is much easier to find a problem if it is 2000 lines PR.

fysnet commented:
I now understand the desire to have them separated. Please forgive me on that. I am just learning this github PR processes, so forgive my ignorance on this subject. I have updated the scsi_device files. I also added a change to the win32res.rc file with the intent of adding a new PR. However it looks like it added it to the current PR. Sorry. Again, just learning.

stlintel commented:
Merged to main. Please continue with new PR if you have more changes planned.

stlintel commented:
Will this PR fix (0xc0000417 crash while exiting with uhci enabled https://github.com/bochs-emu/Bochs/issues/25) ?
